groovy-language-server/src/main/java/com/palantir/groovylanguageserver/GroovycWrapper.java

```diff
@@ -116,6 +131,61 @@ public Path getWorkspaceRoot() {
     }
 
     @Override
+    public Map<String, Set<SymbolInformation>> getReferences() {
+        return references;
+    }
+
+    @Override
+    public Set<Location> findReferences(TextDocumentPositionParams params) {
+        Set<SymbolInformation> symbols = fileSymbols.get(params.getTextDocument().getUri());
+        if (symbols == null) {
+            return Sets.newHashSet();
+        }
+        Set<SymbolInformation> foundSymbolInformations = symbols.stream()
+                .filter(s -> Ranges.isValid(s.getLocation().getRange())
+                        && Ranges.contains(s.getLocation().getRange(), params.getPosition())
+                        && (s.getKind() == SymbolKind.Class || s.getKind() == SymbolKind.Interface))
+                .collect(Collectors.toSet());
+
+        if (foundSymbolInformations.isEmpty()) {
+            return Sets.newHashSet();
+        }
+
+        Optional<SymbolInformation> foundSymbol;
+        if (foundSymbolInformations.size() == 1) {
+            foundSymbol = Optional.of(Iterables.getOnlyElement(foundSymbolInformations));
+        } else {
+            foundSymbol = reduceFoundSymbolsToMostSpecific(foundSymbolInformations, params.getPosition());
+        }
+
+        if (!foundSymbol.isPresent()) {
+            return Sets.newHashSet();
+        }
+
+        String symbolName = foundSymbol.get().getName();
+        if (references.get(symbolName) != null) {
+            return references.get(symbolName).stream().map(s -> s.getLocation())
+                    .filter(l -> Ranges.isValid(l.getRange())).collect(Collectors.toSet());
+        }
+
+        return Sets.newHashSet();
+    }
+
+    private Optional<SymbolInformation> reduceFoundSymbolsToMostSpecific(Set<SymbolInformation> symbols,
```

Comment:
I think it would be cleaner to make this comparison right away in `findReferences` the first time we check if the range contains the position. It'll simplify the code too!